### ISSUE TITLE
Support per-layer rank specification for LoRA adapters

### ISSUE THREAD

**Issue by nshkrdotcom** [thinking-machines-lab/tinker-feedback#26]
## Summary
Enable independent rank configuration for different layer groups within a model, allowing researchers to allocate parameter budget strategically based on empirical sensitivity analysis rather than applying uniform rank across all layers.

## Motivation and Use Case

Current LoRA implementations apply a single rank value uniformly across all adapted layers. However, research in domain-specific fine-tuning demonstrates that different layers exhibit vastly different sensitivities to task-specific adaptations. For narrative synthesis tasks specifically, empirical evidence suggests that mid-layer attention mechanisms (handling semantics and narrative structure) require substantially higher expressiveness than early layers (syntax and grammar) or late layers (vocabulary distribution).

The Jacobian analysis framework shows that gradient norms serve as a proxy for layer sensitivity. When training with uniform rank, researchers observe that mid-attention layers consistently show gradient norms above ten, indicating the model explores high-dimensional spaces, while early and late layers show norms below one, suggesting the rank budget is wasted on regions where the model is rigid. This mismatch results in either under-fitting the critical semantic layers when using conservative global ranks, or over-parameterizing rigid layers when using aggressive ranks.

## Proposed Solution

Extend the LoRA training configuration to accept hierarchical rank specifications:

```python
lora_config = {
    "layers": {
        "0-8": {"rank": 16, "alpha": 32},      # Early attention
        "9-16": {"rank": 64, "alpha": 128},    # Mid attention (narrative structure)
        "17-24": {"rank": 32, "alpha": 64},    # Late attention
        "mlp": {"rank": 32, "alpha": 64}       # MLP layers (optional override)
    }
}
```

This configuration would allow researchers to concentrate parameter budget where narrative coherence actually emerges, potentially achieving better quality per parameter while reducing total parameter counts.

## Technical Considerations

This feature extends existing LoRA infrastructure rather than requiring fundamental architectural changes. The training system already manages per-layer adapter matrices. The enhancement involves exposing more granular control over rank initialization and allowing heterogeneous rank specifications within a single training job.

Key implementation aspects include maintaining compatibility with existing uniform rank configurations, ensuring gradient accumulation and optimizer state management handle varying rank sizes correctly, and providing clear documentation about how rank differences affect checkpoint sizes and inference costs.

## Benefits to Broader Community

This capability addresses a common pain point across multiple research domains beyond narrative synthesis. Researchers working on code generation, mathematical reasoning, and scientific literature processing have reported similar findings about differential layer sensitivity. The ability to allocate rank budgets strategically would enable:

- More efficient use of limited parameter budgets in resource-constrained environments
- Better model quality through targeted capacity allocation
- Clearer experimental protocols for studying where task-specific knowledge actually resides in transformers
- Reduced training costs by avoiding over-parameterization of insensitive layers

## Alternatives Considered

Researchers currently work around this limitation by training multiple separate adapters for different layer groups and composing them, but this approach introduces significant complexity in checkpoint management and inference pipelines. Others simply accept the inefficiency of uniform rank allocation. Neither alternative provides a clean solution.

## Priority Rationale

This represents a high-value, reasonable-complexity enhancement. It extends existing functionality in a natural way, benefits a broad research community beyond a single project, and aligns with Tinker's stated goal of providing researchers with fine-grained control over training dynamics. The feature does not require exposing low-level implementation details or breaking existing abstractions.
